tools/sigclient/pkg/query/validation.go

```diff
@@ -0,0 +1,664 @@
+		}
+	}
+
+	return nil
+}
+
+func sortRecords(records []map[string]interface{}, columns []string) {
+
+	sort.Slice(records, func(i, j int) bool {
+		for _, column := range columns {
+			valI, valJ := records[i][column], records[j][column]
+
+			strI := fmt.Sprintf("%v", valI)
+			strJ := fmt.Sprintf("%v", valJ)
+
+			if strI == strJ {
+				continue
+			}
+
+			return strI < strJ
+		}
+		return false
+	})
+}
+
+/*
+*
+  - ValidateLogsQueryResults validates the logs query results
+    Compares:
+  - TotalMatched
+  - AllColumns
+  - ColumnsOrder
+  - Records
+  - @param queryRes: Query results from the server
+  - @param expRes: Expected results
+  - @return error: Error if validation fails
+
+*
+*/
+func ValidateLogsQueryResults(queryRes *Result, expRes *Result) error {
+	var err error
+
+	equal := reflect.DeepEqual(queryRes.TotalMatched, expRes.TotalMatched)
+	if !equal {
+		return fmt.Errorf("ValidateLogsQueryResults: TotalMatched mismatch, expected: %v, got: %v", expRes.TotalMatched, queryRes.TotalMatched)
+	}
+	colsToRemove := map[string]struct{}{
+		"timestamp": {},
+		"_index":    {},
+	}
+
+	queryRes.AllColumns = utils.RemoveValues(queryRes.AllColumns, colsToRemove)
+	expRes.AllColumns = utils.RemoveValues(expRes.AllColumns, colsToRemove)
+
+	equal = utils.ElementsMatch(queryRes.AllColumns, expRes.AllColumns)
+	if !equal {
+		return fmt.Errorf("ValidateLogsQueryResults: AllColumns mismatch, expected: %+v, got: %+v", expRes.AllColumns, queryRes.AllColumns)
+	}
+
+	queryRes.ColumnsOrder = utils.RemoveValues(queryRes.ColumnsOrder, colsToRemove)
+	expRes.ColumnsOrder = utils.RemoveValues(expRes.ColumnsOrder, colsToRemove)
+
+	equal = utils.CompareSlices(queryRes.ColumnsOrder, expRes.ColumnsOrder)
+	if !equal {
+		return fmt.Errorf("ValidateLogsQueryResults: ColumnsOrder mismatch, expected: %+v, got: %+v", expRes.ColumnsOrder, queryRes.ColumnsOrder)
+	}
+	if len(queryRes.Records) < len(expRes.Records) {
+		return fmt.Errorf("ValidateLogsQueryResults: Less records than, expected at least: %v, got: %v", len(expRes.Records), len(queryRes.Records))
+	}
+
+	queryRes.UniqueKeyCols = expRes.UniqueKeyCols
+	err = ValidateUniqueKeyColsInResult(queryRes)
+	if err != nil {
+		return fmt.Errorf("ValidateLogsQueryResults: Error validating UniqueKeyCols: %v in queryRes, err: %v", queryRes.UniqueKeyCols, err)
+	}
+
+	sortRecords(queryRes.Records, queryRes.UniqueKeyCols)
+	sortRecords(expRes.Records, expRes.UniqueKeyCols)
+
+	for idx, record := range expRes.Records {
+		err = ValidateRecord(queryRes.Records[idx], record)
+		if err != nil {
+			return fmt.Errorf("Error comparing records: queryRes Record: %v, expRes Record: %v, err: %v", queryRes.Records[idx], record, err)
+		}
+	}
+
+	return nil
+}
+
+func getGroupByCombination(grpByValues []string) string {
+	grpByValue := ""
+	for _, value := range grpByValues {
+		grpByValue += value + "_"
+	}
+	return grpByValue
+}
```

Comment:
This is going to have a trailing `_`; is that what you want? If you don't (of if it doesn't matter), you can use `strings.Join()` instead of making this helper function

Reply:
addressed in the latest commit.